ibc/docs/ics23.md

```diff
@@ -0,0 +1,78 @@
+
+As the SMT used by Pocket Network only stores hashed values by default, the IBC stores use the `WithValueHasher(nil)` option which stores the raw bytes of the values in the tree. As such the following `ProofSpec` was created:
+
+```go
+smtSpec *ics23.ProofSpec = &ics23.ProofSpec{
+	LeafSpec: &ics23.LeafOp{
+		Hash:         ics23.HashOp_SHA256,
+		PrehashKey:   ics23.HashOp_SHA256,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+        Prefix:       []byte{0},
+	},
+	InnerSpec: &ics23.InnerSpec{
+		ChildOrder:      []int32{0, 1},
+		ChildSize:       32,
+		MinPrefixLength: 1,
+		MaxPrefixLength: 1,
+        EmptyChild:      make([]byte, 32),
+		Hash:            ics23.HashOp_SHA256,
+	},
+	MaxDepth:                   256,
+	PrehashKeyBeforeComparison: true,
+}
+```
+
+The main change from the `cosmos/ics23` `SmtSpec` object is that the `PrehashValue` field is set to not hash values prior to hashing the key-value pair.
+
+### Converting `SparseMerkleProof` to `CommitmentProof`
+
+In order to convert the proofs generated by the SMT into a serialisable proof used by `cosmos/ics23`, the `SideNodes` field of the `SparseMerkleProof` must be converted into a list of `InnerOp` types which define the order of the hashes. The order of the hashes is important as depending on whether the next hash is the left or right neighbour of the current hash, they will be hashed in a different order, ultimately creating a different root hash. This conversion allows the verification to produce the same root hash as the SMT would have produced when verifying the proof.
+
+As `SparseMerkleProof` objects represent both inclusion and exclusion proofs as defined in the [JMT whitepaper][jmt] the conversion step will convert the SMT proof into either a `ExistenceProof` or `ExclusionProof` as defined in `cosmos/ics23`.
+
+### Proof Verification
+
+When verifying membership of an element the logic is as follows:
+
+1. Use the key-value pair to generate a leaf hash
+2. Hash the leaf with the `SideNodes` found in the `path` field of the `ExistenceProof` to generate the root hash
+3. Compare the root hash with the one provided
+
+For non-membership the logic is as follows:
+
+1. If the `ActualValueHash` field in the `ExclusionProof` is the SMT's placeholder value, then use the placeholder value as the leaf node hash and follow steps 3-4
+2. If the `ActualValueHash` field is not the placeholder value, then use the `ActualPath` and `ActualValueHash` fields to generate the leaf node hash - do not hash these values before hashing the node as they are populated from the SMT proof's `NonMembershipLeafData` field and thus are already hashed
+3. Hash the leaf node hash with the `SideNodes` found in the `Path` field of the `ExclusionProof` to generate the root hash
+4. Compare the root hash with the one provided
```

Comment:
because we're using `ActualValueHash`, we also verify that the root hash is the same as the one computed even though it's different from the value the user gives?

Can you clarify this point